File: Daily/2130.maximum-twin-sum-of-a-linked-list.js
/*
 * @lc app=leetcode id=2130 lang=javascript
 *
 * [2130] Maximum Twin Sum of a Linked List
 */

function ListNode(val, next) {
    this.val = val === undefined ? 0 : val;
    this.next = next === undefined ? null : next;
}

// @lc code=start
/**
 * @param {ListNode} head
 * @return {number}
 */
var pairSum = function (head) {
    /** @type {number[]} */
    const pairSums = [];

    /** @type {ListNode | null} */
    let fast = head;
    let slow = head;

    while (true) {
        fast = fast.next;
        fast = fast.next;
        pairSums.push(slow.val);
        if (fast === null) break;
        slow = slow.next;
    }

    slow.next = reverseLinkedList(slow.next);

    let maxPairSum = 0;

    let firstHalfPointer = head;
    let secondHalfPointer = slow.next;

    while (secondHalfPointer !== null) {
        maxPairSum = Math.max(
            maxPairSum,
            firstHalfPointer.val + secondHalfPointer.val,
        );

        firstHalfPointer = firstHalfPointer.next;
        secondHalfPointer = secondHalfPointer.next;
    }

    // slow.next = reverseLinkedList(slow.next);

    return maxPairSum;
};

/**
 * @param {ListNode | null} head
 * @returns {ListNode | null}
 */
function reverseLinkedList(head) {
    if (head === null || head.next === null) return head;

    const fakeHead = new ListNode();

    /** @type {ListNode | null} */
    let current = head;
    /** @type {ListNode | null} */
    let next = head.next;

    while (current !== null) {
        next = current.next;

        const newNext = fakeHead.next;
        current.next = newNext;
        fakeHead.next = current;

        current = next;
    }

    return fakeHead.next;
}
// @lc code=end

File: src/2026/2130.maximum-twin-sum-of-a-linked-list.cpp
/*
 * @lc app=leetcode id=2130 lang=cpp
 *
 * [2130] Maximum Twin Sum of a Linked List
 */

#include <algorithm>
#include <memory>

struct ListNode {
  int val;
  ListNode* next;
  ListNode() : val(0), next(nullptr) {}
  ListNode(int x) : val(x), next(nullptr) {}
  ListNode(int x, ListNode* next) : val(x), next(next) {}
};

// @lc code=start
class Solution {
 public:
  int pairSum(ListNode* head) {
    auto fake_head = std::make_unique<ListNode>();
    fake_head->next = head;
    ListNode* p1 = fake_head.get();
    ListNode* p2 = fake_head.get();
    while (p2->next != nullptr) {
      p1 = p1->next;
      p2 = p2->next->next;
    }
    // Since list always has even nodes,
    // now p1 is the prev node of second half head
    p1->next = Reverse(p1->next);

    ListNode* first_half_ptr = head;
    ListNode* second_half_ptr = p1->next;

    int max_pair_sum = 0;
    while (second_half_ptr != nullptr) {
      max_pair_sum =
          std::max(max_pair_sum, first_half_ptr->val + second_half_ptr->val);
      first_half_ptr = first_half_ptr->next;
      second_half_ptr = second_half_ptr->next;
    }
    return max_pair_sum;
  }

 private:
  static ListNode* Reverse(ListNode* head) {
    auto fake_head = std::make_unique<ListNode>();

    ListNode* current_node = head;
    while (current_node != nullptr) {
      ListNode* next_node = current_node->next;

      current_node->next = fake_head->next;
      fake_head->next = current_node;

      current_node = next_node;
    }

    return fake_head->next;
  }
};
// @lc code=end
